keps/sig-node/asier/README.md

```diff
@@ -0,0 +1,530 @@
+
+  
+
+### Goals
+
+  
+
+* Allow IMA to work inside a container with remote attestation
+
+  
+
+### Non-Goals
+
+  
+
+## Proposal
+  
+We propose to enable IMA linux namespaces in pods.
+
+Since IMA namespaces can be created when a container is launched, we can provide transparent integrity verification on any linux container.
+
+IMA and EVM can use a TPM chip as a hardware root of trust. Hence we can verify images against a set of golden hash values, as well as avoiding any further changes to the overlayfs to intercept calls and check the integrity of files.
+
+
+### User Stories (Optional)
+
+  
+
+#### Story 1
+
+As a cluster admin, I want to detect undesired file changes, so that I can take out pods that have been compromised.
+
+  
+
+#### Story 2
+
+  
+As a cluster admin, I want to deploy a only proven and certified pods, so that I can comply with internal policies as well as security regulations.
+
+#### Story 3
+
+  
+As a cluster admin, I want to deny access to certain files inside the pod, so that a potential intruder can't access sensitive information.
+
+### Notes/Constraints/Caveats (Optional)
+
+  
+
+We need to enable IMA in the kernel and container runtimes (runC, CRI-O, docker, containerd, etc.). 
+  
+
+### Risks and Mitigations
+
+  
+
+## Design Details
+
+  
+
+<!--
+
+This section should contain enough information that the specifics of your
+
+change are understandable. This may include API specs (though not always
+
+required) or even code snippets. If there's any ambiguity about HOW your
+
+proposal will be implemented, this is the place to discuss them.
+
+-->
+
+In order for the this feature to work, the nodes where the pods will be deploy should have IMA enabled and a recent kernel that supports IMA namespaces (WIP, it should be merged soon). The nodes should also have a TPM chip. We could use nodAffinity of labels and annotations in nodes in order to select where to deploy the pods.
+
+The linux kernel IMA namespace support is based on user namespaces. Therefore, the container runtime should first create a user namespace and then create an IMA namespaces. In order to use IMA namespaces it is necessary to enable user namespaces as well.
+
+Should we enable IMA namespaces by default when enabling user namespaces?
+
+
+There will be a CRI API change which will allow the pod to use IMA namespaces and specify the namespace policy.
+
+  
+
+### Linux kernel
+
+  
+
+IMA is only available in Linux hosts and Linux containers. Unfortunately, IMA is not a separate namespace, which is needed in order to isolate it and be used inside containers. Upcoming kernel patches should add support for IMA namespaces.
+
+  
+
+### Runtime specification
+
+In order to run a container we need a bundle that contains a config.json file with all the configuration. There is a Linux specific configuration section where the namespaces are listed. According to the standard, the following namespace types SHOULD be supported.
```

Comment:
You shouldn't explain the runtime-spec change here. At most mention that a PR is merged/open for this, but not go into the details here in the KEP. Lot of people will comment things that will be lost with the upstream discussion for runtime-spec.

Let's keep the runtime-spec discussion in the runtime-spec PR and remove most of this from here.

Reply:
Thanks, I will link to a PR in runC that shows how it can be achieved.